src/Sentry.DiagnosticSource/Internal/DiagnosticSource/SentryEFCoreListener.cs

```diff
@@ -52,85 +45,54 @@ public SentryEFCoreListener(IHub hub, SentryOptions options)
-        if (transaction == null)
-        {
-            return;
-        }
-
-        var parent = type == SentryEFSpanType.QueryExecution
-            ? transaction.GetLastActiveSpan() ?? transaction.GetDbParentSpan()
-            : transaction.GetDbParentSpan();
-
-        var child = parent.StartChild(operation, description);
-
-        var asyncLocalSpan = GetSpanBucket(type);
-        asyncLocalSpan.Value = new WeakReference<ISpan>(child);
-    }
-
-    private ISpan? TakeSpan(SentryEFSpanType type)
-    {
-        var transaction = _hub.GetTransactionIfSampled();
-        if (transaction == null)
-        {
-            return null;
-        }
-
-        if (GetSpanBucket(type).Value is { } reference && reference.TryGetTarget(out var startedSpan))
-        {
-            return startedSpan;
-        }
+    public void OnCompleted() { }
 
-        _options.LogWarning("Trying to close a span that was already garbage collected. {0}", type);
-        return null;
-    }
+    public void OnError(Exception error) { }
 
-    private AsyncLocal<WeakReference<ISpan>> GetSpanBucket(SentryEFSpanType type)
-        => type switch
-        {
-            SentryEFSpanType.QueryCompiler => _spansCompilerLocal,
-            SentryEFSpanType.QueryExecution => _spansQueryLocal,
-            SentryEFSpanType.Connection => _spansConnectionLocal,
-            _ => throw new($"Unknown SentryEFSpanType: {type}")
-        };
+    private EFQueryCompilerDiagnosticSourceHelper QueryCompilerDiagnosticSourceHelper(object? diagnosticsSourceValue) =>
+        new(_hub, _options, _spansCompilerLocal, diagnosticsSourceValue);
 
-    public void OnCompleted() { }
+    private EFConnectionDiagnosticSourceHelper ConnectionDiagnosticSourceHelper(object? diagnosticsSourceValue) =>
+        new(_hub, _options, _spansConnectionLocal, diagnosticsSourceValue);
 
-    public void OnError(Exception error) { }
+    private EFCommandDiagnosticSourceHelper CommandDiagnosticSourceHelper(object? diagnosticsSourceValue) =>
+        new(_hub, _options, _spansQueryLocal, diagnosticsSourceValue);
 
     public void OnNext(KeyValuePair<string, object?> value)
     {
         try
         {
+            // Because we have to support the .NET framework, we can't get at strongly typed diagnostic source events.
+            // We do know they're objects, that can be converted to strings though... and we can get the correlation
+            // data we need from there by parsing the string. Not as reliable, but works against all frameworks.
+            var diagnosticSourceValue = value.Value?.ToString();
             switch (value.Key)
             {
                 // Query compiler span
                 case EFQueryStartCompiling or EFQueryCompiling:
-                    AddSpan(SentryEFSpanType.QueryCompiler, "db.query.compile", FilterNewLineValue(value.Value));
+                    QueryCompilerDiagnosticSourceHelper(value.Value).AddSpan();
```

Comment:
These factory methods allocate new instances of the helper classes every time they are called.  That's going to be a lot of allocations under load.  Please refactor to either use static helpers, or a single shared instance.

I don't think there is state shared in the helper, because we're getting a different helper instance for each diagnostic event.  Right?